src/cfg/cfg-traversal.h

```diff
@@ -205,47 +211,79 @@ struct CFGWalker : public ControlFlowWalker<SubType, VisitorType> {
     self->startUnreachableBlock();
   }
 
-  static void doEndCall(SubType* self, Expression** currp) {
+  static void doEndThrowingInst(SubType* self, Expression** currp) {
     // Every call can possibly throw, but we don't end the current basic block
     // unless the call is within a try-catch, because the CFG will have too many
     // blocks that way, and if an exception is thrown, the function will be
     // exited anyway.
-    if (!self->catchStack.empty()) {
-      auto* last = self->currBasicBlock;
-      self->startBasicBlock();
-      self->link(last, self->currBasicBlock);    // exception not thrown
-      self->link(last, self->catchStack.back()); // exception thrown
+    if (!self->unwindCatchStack.empty()) {
+      // Exception thrown. Create a link to each catch within the innermost try.
+      for (auto* block : self->unwindCatchStack.back()) {
+        self->link(self->currBasicBlock, block);
+      }
+      // If the innermost try does not have a catch_all clause, an exception
+      // thrown can be caught by any of its outer catch block. And if that outer
+      // try-catch also does not have a catch_all, this continues until we
+      // encounter try-catch_all. Create a link to all those possible catch
+      // unwind destinations.
```

Comment:
Is it necessary to keep track of event types here so that a link is only created to the innermost catch block that handles each event type? AFAIK that would be more precise, but I don't know how precise this CFG needs to be. If it's not necessary now, it might be worth mentioning in a TODO if it would help with any optimizations later.

Reply:
I think we can do something like that for `throw`s, but not calls, because we don't know what event type will be thrown from calls. And given that the vast majority of throwing instructions will be calls, I'm not sure how much impact that can have, but I'll add a TODO here:
```
// TODO This can be more precise for `throw`s if we compare event types    
// and create links to outer catch BBs only when the exception is not      
// caught.
```

Reply:
I think this applies for calls, too. For example:

```
try
  try
    call $foo
  catch $e1
    ...
  catch $e2
    ...
  end
catch $e1
  ...
catch $e3
  ...
end
```
Here the call to $foo can go to either of the inner catch blocks or the outer catch of `$e3`, but it will never arrive directly in the outer catch of `$e1`.

Reply:
Thanks, that can be a good optimization opportunity. Do you mind if I mark that as a TODO for now?

Reply:
Added the following TODO comment:
```
    // TODO This can also be more precise if we analyze the structure of nested
    // try-catches. For example, in the example below, 'call $foo' doesn't need
    // a link to the BB of outer 'catch $e1', because if the exception thrown by
    // the call is of event $e1, it would've already been caught by the inner
    // 'catch $e1'. Optimize these cases later.
    // try
    //   try
    //     call $foo
    //   catch $e1
    //     ...
    //   catch $e2
    //     ...
    //   end
    // catch $e1
    //   ...
    // catch $e3
    //   ...
    // end
```

Reply:
Sounds good!